### ISSUE TITLE
O_CLOEXEC requires a higher _XOPEN_SOURCE value

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nix#1091]
From [here](http://man7.org/linux/man-pages/man2/open.2.html):

> The O_CLOEXEC, O_DIRECTORY, and O_NOFOLLOW flags are not specified in
       POSIX.1-2001, but are specified in POSIX.1-2008.  Since glibc 2.12,
       one can obtain their definitions by defining either _POSIX_C_SOURCE
       with a value greater than or equal to 200809L or _XOPEN_SOURCE with a
       value greater than or equal to 700.  In glibc 2.11 and earlier, one
       obtains the definitions by defining _GNU_SOURCE.

At the same time, in [archive.cc](https://github.com/NixOS/nix/blob/master/src/libutil/archive.cc#L1), we have:

```c
#define _XOPEN_SOURCE 600
```

followed [farther down the file](https://github.com/NixOS/nix/blob/master/src/libutil/archive.cc#L44) with:

```c
    AutoCloseFD fd = open(path.c_str(), O_RDONLY | O_CLOEXEC);
```

I think the darwin build stack is stricter than the GNU one about this sort of thing, and that's what's causing [this bug](https://github.com/NixOS/nix/issues/1013).

cc @edolstra @shlevy

**Comment by shlevy** [comment]
:+1: does it all work with `700`?

**Comment by copumpkin** [comment]
I haven't tried yet, but it seems like it's explicitly going against spec right now, so it seems like a no regrets fix. I also think @mstone tried a few things like this on his branch.

**Comment by shlevy** [comment]
Hmm we only define XOPEN_SOURCE in that one dir but we use cloexec lots of places :frowning_face:

**Comment by shlevy** [comment]
@edolstra Why is that define even there?

**Comment by copumpkin** [comment]
I think it's more that if you do define `_XOPEN_SOURCE`, it suggests you want a particular version of the standard. If not, I think platforms tend to default to fairly recent specs. I'll experiment a bit and see what the behavior is on Darwin.

**Comment by copumpkin** [comment]
Fixed by 19c278de89fa60844fa1ea2b3c2abfa639f75b39
